File: agent/java/com/squeeze/agent/net/TeeStreams.java
package com.squeeze.agent.net;

import java.io.FilterInputStream;
import java.io.FilterOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/** Stream wrappers that copy bytes into a SqueezeTracker as they flow. */
final class TeeStreams {

    static final class RequestStream extends FilterOutputStream {
        private final SqueezeTracker tracker;
        RequestStream(OutputStream out, SqueezeTracker t) { super(out); this.tracker = t; }
        @Override public void write(int b) throws IOException {
            out.write(b);
            tracker.appendRequestBody(new byte[]{(byte) b}, 0, 1);
        }
        @Override public void write(byte[] b, int off, int len) throws IOException {
            out.write(b, off, len);
            tracker.appendRequestBody(b, off, len);
        }
    }

    static final class ResponseStream extends FilterInputStream {
        private final SqueezeTracker tracker;
        ResponseStream(InputStream in, SqueezeTracker t) { super(in); this.tracker = t; }
        @Override public int read() throws IOException {
            int b = in.read();
            if (b >= 0) tracker.appendResponseBody(b);
            else tracker.report();
            return b;
        }
        @Override public int read(byte[] b, int off, int len) throws IOException {
            int n = in.read(b, off, len);
            if (n > 0) tracker.appendResponseBody(b, off, n);
            else if (n < 0) tracker.report();
            return n;
        }
        @Override public void close() throws IOException {
            tracker.report();
            super.close();
        }
    }

    private TeeStreams() {}
}

File: agent/kotlin/com/squeeze/capture/TeeStreams.kt
package com.squeeze.capture

import java.io.FilterInputStream
import java.io.FilterOutputStream
import java.io.InputStream
import java.io.OutputStream

/** Output stream that copies written bytes into the tracker as the request body. */
class RequestStream(out: OutputStream, private val tracker: SqueezeTracker) : FilterOutputStream(out) {
    override fun write(b: Int) {
        out.write(b)
        tracker.appendRequestBody(byteArrayOf(b.toByte()), 0, 1)
    }
    override fun write(b: ByteArray, off: Int, len: Int) {
        out.write(b, off, len)
        tracker.appendRequestBody(b, off, len)
    }
}

/** Input stream that copies read bytes into the tracker as the response body. */
class ResponseStream(input: InputStream, private val tracker: SqueezeTracker) : FilterInputStream(input) {
    override fun read(): Int {
        val b = `in`.read()
        if (b >= 0) tracker.appendResponseByte(b) else tracker.report()
        return b
    }
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        val n = `in`.read(b, off, len)
        if (n > 0) tracker.appendResponseBody(b, off, n) else if (n < 0) tracker.report()
        return n
    }
    override fun close() {
        tracker.report()
        super.close()
    }
}
